README.md
# Marvin
Marvin is the ultimate tool to visualise and analyse MaNGA data. It is developed and maintained by the MaNGA team.

[![Build Status](https://travis-ci.org/sdss/marvin.svg?branch=master)](https://travis-ci.org/sdss/marvin)
[![Coverage Status](https://coveralls.io/repos/github/sdss/marvin/badge.svg?branch=master)](https://coveralls.io/github/sdss/marvin?branch=master)
[![PyPI](https://img.shields.io/pypi/v/sdss-marvin.svg)](https://pypi.python.org/pypi/sdss-marvin)
[![PyPI download](https://img.shields.io/pypi/dm/sdss-marvin.svg)](https://pypi.python.org/pypi/sdss-marvin)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.292632.svg)](https://doi.org/10.5281/zenodo.292632)
[![astropy](http://img.shields.io/badge/powered%20by-AstroPy-orange.svg?style=flat)](http://www.astropy.org/)
![BrowserStack Status](https://www.browserstack.com/automate/badge.svg?badge_key=WWgyaGJBbW45aityUVJtYytDcHFydU9EZE9ObVdOVVpkaUxGZkxwbzdHQT0tLUNkcW5Hc3JaRTdqR0l6ajltSUdTRXc9PQ==--21b221b6714b852f8f4215c787ffa6e2812e2ad6)
[![readthedocs](https://readthedocs.org/projects/docs/badge/)](http://sdss-marvin.readthedocs.io/en/latest/)


Installation
------------

To painlessly install Marvin:

    pip install sdss-marvin

If you would like to contribute to Marvin's development, you can clone this git repo, pip install the dependencies, update the submodules, and then setup with `python setup.py install`:

    git clone https://github.com/sdss/marvin
    cd marvin
    pip install -r requirements_dev.txt
    git submodule init
    git submodule update
    python setup.py install


What is Marvin?
---------------

Marvin is a complete ecosystem designed for overcoming the challenge of
searching, accessing, and visualizing the MaNGA data. It consists of a
three-pronged approach of a web app, a python package, and an API. The web app,
Marvin-web, provides an easily accessible interface for searching the MaNGA data
and visual exploration of individual MaNGA galaxies or of the entire sample. The
python package, in particular Marvin-tools, allows users to easily and
efficiently interact with the MaNGA data via local files, files retrieved from
the [Science Archive Server](https://sas.sdss.org), or data directly grabbed
from the database.  The tools come mainly in the form of convenience functions
and classes for interacting with the data. An additional tool is a powerful
query functionality that uses the API to query the MaNGA databases and return
the search results to your python session. Marvin-API is the critical link that
allows Marvin-tools and Marvin-web to interact with the databases, which enables
users to harness the statistical power of the MaNGA data set.

Documentation
-------------

You can find the latest Marvin documentation [here](http://sdss-marvin.readthedocs.io/en/latest/).


Citation and Acknowledgements
-----------------------------

If you use Marvin for work/research presented in a publication (whether directly, or as a dependency to another package), we ask that you cite the [Marvin Software](https://zenodo.org/record/292632) (BibTeX). We provide the following as a standard acknowledgment you can use if there is not a specific place to cite the DOI:

    *This research made use of Marvin, a core Python package and web framework for MaNGA data, developed by Brian Cherinka,
    José Sánchez-Gallego, and Brett Andrews. (MaNGA Collaboration, 2016).*

Marvin's Bibtex entry to use:

      @misc{brian_cherinka_2017_292632,
        author       = {Brian Cherinka and
                        José Sánchez-Gallego and
                        Brett Andrews},
        title        = {sdss/marvin: Marvin Beta 2.1.0},
        month        = feb,
        year         = 2017,
        doi          = {10.5281/zenodo.292632},
        url          = {https://doi.org/10.5281/zenodo.292632}
      }

License
-------
Marvin is licensed under a 3-clause BSD style license - see the
``LICENSE.md`` file.


CHANGELOG.md
# Marvin's Change Log

## [2.1.5] - unreleased
### Added:
- Better BPT documentation, in particular in the `Modifying the plot` section.
- A hack function ``marvin.utils.plot.utils.bind_to_figure()`` that replicate the contents of a matplotlib axes in another figure.

### Changed:
- Issue #190: ``Maps.get_bpt()`` and ``marvin.utils.dap.bpt.bpt_kewley06()`` now also return a list of axes. Each axes contains a method pointing to the ``marvin.utils.plot.utils.bind_to_figure()`` function, for easily transfer the axes to a new figure.

### Fixed:


## [2.1.4] - 2017/08/02
### Added:
- Added new query_params object, for easier navigation of available query parameters.  Added new tests.
- Added a new guided query builder using Jquery Query Builder to the Search page
- Added a View Galaxies link on the web results to view postage stamps of the galaxies in the results
- Added Route Rate Limiting.  Adopts a limit of 200/min for all api routes and 60/minute for query api calls and web searches

### Changed:
- Changed call signature for :meth:`marvin.utils.plot.map.no_coverage_mask` (removed ``value`` arg because unused, added ``None`` as default value ``ivar`` (``None``), and re-ordered args and kwargs).
- Changed call signature for :meth:`marvin.utils.plot.map.bad_data_mask` (removed ``value`` arg because unused).
- Changed the Marvin web search page to use the new query_params and parameter grouping.  Removed the autocomplete input box.
- Updated the documentation on query and query_params.
- Modified Guided Search operator options to remove options that could not be parsed by SQLA boolean_search
- Refactored the web settings, route registration, extensions to enable extensibility
- Issue #282: Improvements to "Go to CAS" link.  Changed to Go To SkyServer and updated link to public up-to-date link

### Fixed:
- Issue #102: problem with urllib package when attempting to retrieve the Marvin URLMap
- Issue #93: safari browser does not play well with marvin
- Issue #155: Contrails in Web Map
- Issue #174: sdss_access may not be completely python 3 compatible
- Issue #196: Bin not loading from local sas
- Issue #207: Get Maps in MapSpecView of Galaxy page sometimes fails to return selected maps
- Issue #210: pip upgrade may not install new things as fresh install
- Issue #209: marvin version from pip install is incorrect
- Issue #268: Cube flux from file error
- Issue #85: Python does not start in Python 3
- Issue #273: ha.value bug
- Issue #277: Ticks for log normalized colorbar
- Issue #275: logger crashes on warning when other loggers try to log
- Issue #258: 422 Invalid Parameters
- Issue #271: Problem in dowloading image.
- Issue #97: sqlalchemy-boolean-search not found when installed from pip source
- Issue #227: Marvin installation in python 3.6 (update setuptools to 36)
- Issue #262: problem with marvin update
- Issue #270: BPT array sizing not compatible
- Issue #88: Deployment at Utah requires automatisation
- Issue #234: Add (and use) functions to the datamodel to determine plotting parameters
- Issue #278: marvin_test_if decorator breaks in python 2.7
- Issue #274: cube slicing to get a spaxel fails with maps error
- Issue #39: implement more complete testing framework
- Issue #242: Result object representation error with 0 query results
- Issue #159: Marvin issues multiple warnings in PY3
- Issue #149: Improve integrated flux maps display in web

## [2.1.3] - 2017/05/18
### Added:
- Issue #204: added elpetro_absmag colours to mangaSampleDB models.
- Issue #253: Plotting tutorial.
- Issue #223: Easy multi-panel map plotting (with correctly placed colorbars).
- Issue #232 and Issue #251: Uses matplotlib style sheets context managers for plotting (map, spectrum, and BPT) and restores previous defaults before methods finish.
- Issue #189: Map plotting accepts user-defined value, ivar, and/or mask (including BPT masks).
- Issue #252: Quantile clipping for properties other than velocity, sigma, or flux in web.
- Added `utils.plot.map` doc page.
- Added `tools.map` doc page.

### Changed:
- Issue #243: inverted `__getitem__` behaviour for Cube/Maps/ModelCube and fixed tests.
- Modified Flask Profiler File to always point to $MARVIN_DIR/flask_profiler.sql
- Issue #241: Moved map plotting methods from tools/map to utils/plot/map
- Issue #229 and Issue #231: Switch to new gray/hatching scheme (in tools and web):
  - gray: spaxels with NOCOV.
  - hatched: spaxels with bad data (UNRELIABLE and DONOTUSE) or S/N below some minimum value.
  - colored: good data.
- Issue #238: Move plot defaults to datamodel (i.e., bitmasks, colormaps, percentile clips, symmetric, minimum SNR).
- Issue #206: SNR minimum to None (effectively 0) for velocity maps so that they aren't hatched near the zero velocity contour.
- Simplified default colormap name to "linearlab."
- Decreased map plot title font size in web so that it does not run onto second line and overlap plot.

### Fixed:
- Interactive prompt for username in sdss_access now works for Python 3.
- Fixed #195: The data file for the default colormap for `Map.plot()` ("linear_Lab") is now included in pip version of Marvin and does not throw invalid `FileNotFoundError` if the data file is missing.
- Fixed #143: prevents access mode to go in to remote if filename is present.
- Fixed #213: shortcuts are now only applied on full words, to avoid blind replacements.
- Fixed #206: no longer masks spaxels close to zero velocity contour in web and tools map plots
- Fixed #229: corrects web bitmask parsing for map plots

python/marvin/api/query.py
from flask_classy import route
from flask import request, jsonify, Response, current_app
from marvin.tools.query import doQuery, Query
from marvin.core.exceptions import MarvinError
from marvin.api.base import BaseView, arg_validate as av
from marvin.utils.db import get_traceback
from marvin.tools.query.query_utils import bestparams
from marvin.web.extensions import limiter
import json


def _run_query(searchfilter, **kwargs):
    ''' Run the query and return the query and results '''

    release = kwargs.pop('release', None)
    kwargs['returnparams'] = kwargs.pop('params', None)
    kwargs['returntype'] = kwargs.pop('rettype', None)

    try:
        q, r = doQuery(searchfilter=searchfilter, release=release, **kwargs)
    except Exception as e:
        raise MarvinError('Query failed with {0}: {1}'.format(e.__class__.__name__, e))
    else:
        return q, r


def _getCubes(searchfilter, **kwargs):
    """Run query locally at Utah and format the output into the full JSON """

    # run the query
    q, r = _run_query(searchfilter, **kwargs)
    results = r.results

    # get the subset keywords
    start = kwargs.get('start', None)
    end = kwargs.get('end', None)
    limit = kwargs.get('limit', None)
    params = kwargs.get('params', None)

    # get a subset
    chunk = None
    if start:
        chunk = int(end) - int(start)
        results = r.getSubset(int(start), limit=chunk)
    chunk = limit if not chunk else limit
    runtime = {'days': q.runtime.days, 'seconds': q.runtime.seconds, 'microseconds': q.runtime.microseconds}
    output = dict(data=results, query=r.showQuery(), chunk=limit,
                  filter=searchfilter, params=q.params, returnparams=params, runtime=runtime,
                  queryparams_order=q.queryparams_order, count=len(results), totalcount=r.totalcount)
    return output


class QueryView(BaseView):
    """Class describing API calls related to queries."""
    decorators = [limiter.limit("60/minute")]

    def index(self):
        '''Returns general query info

        .. :quickref: Query; Get general query info

        :query string release: the release of MaNGA
        :resjson int status: status of response. 1 if good, -1 if bad.
        :resjson string error: error message, null if None
        :resjson json inconfig: json of incoming configuration
        :resjson json utahconfig: json of outcoming configuration
        :resjson string traceback: traceback of an error, null if None
        :resjson string data: data message
        :resheader Content-Type: application/json
        :statuscode 200: no error
        :statuscode 422: invalid input parameters

        **Example request**:

        .. sourcecode:: http

           GET /marvin2/api/query/ HTTP/1.1
           Host: api.sdss.org
           Accept: application/json, */*

        **Example response**:

        .. sourcecode:: http

           HTTP/1.1 200 OK
           Content-Type: application/json
           {
              "status": 1,
              "error": null,
              "inconfig": {"release": "MPL-5"},
              "utahconfig": {"release": "MPL-5", "mode": "local"},
              "traceback": null,
              "data": "this is a query!"
           }

        '''
        self.results['data'] = 'this is a query!'
        self.results['status'] = 1
        return jsonify(self.results)

    @route('/cubes/', methods=['GET', 'POST'], endpoint='querycubes')
    @av.check_args(use_params='query', required='searchfilter')
    def cube_query(self, args):
        ''' Performs a remote query

        .. :quickref: Query; Perform a remote query

        :query string release: the release of MaNGA
        :form searchfilter: your string searchfilter expression
        :form params: the list of return parameters
        :form rettype: the string indicating your Marvin Tool conversion object
        :form limit: the limiting number of results to return for large results
        :form sort: a string parameter name to sort on
        :form order: the order of the sort, either ``desc`` or ``asc``
        :resjson int status: status of response. 1 if good, -1 if bad.
        :resjson string error: error message, null if None
        :resjson json inconfig: json of incoming configuration
        :resjson json utahconfig: json of outcoming configuration
        :resjson string traceback: traceback of an error, null if None
        :resjson string data: dictionary of returned data
        :json list results: the list of results
        :json string query: the raw SQL string of your query
        :json int chunk: the page limit of the results
        :json string filter: the searchfilter used
        :json list returnparams: the list of return parameters
        :json list params: the list of parameters used in the query
        :json list queryparams_order: the list of parameters used in the query
        :json dict runtime: a dictionary of query time (days, minutes, seconds)
        :json int totalcount: the total count of results
        :json int count: the count in the current page of results
        :resheader Content-Type: application/json
        :statuscode 200: no error
        :statuscode 422: invalid input parameters

        **Example request**:

        .. sourcecode:: http

           GET /marvin2/api/query/cubes/ HTTP/1.1
           Host: api.sdss.org
           Accept: application/json, */*

        **Example response**:

        .. sourcecode:: http

           HTTP/1.1 200 OK
           Content-Type: application/json
           {
              "status": 1,
              "error": null,
              "inconfig": {"release": "MPL-5", "searchfilter": "nsa.z<0.1"},
              "utahconfig": {"release": "MPL-5", "mode": "local"},
              "traceback": null,
              "chunk": 100,
              "count": 4,
              "data": [["1-209232",8485,"8485-1901","1901",0.0407447],
                       ["1-209113",8485,"8485-1902","1902",0.0378877],
                       ["1-209191",8485,"8485-12701","12701",0.0234253],
                       ["1-209151",8485,"8485-12702","12702",0.0185246]
                      ],
              "filter": "nsa.z<0.1",
              "params": ["cube.mangaid","cube.plate","cube.plateifu","ifu.name","nsa.z"],
              "query": "SELECT ... FROM ... WHERE ...",
              "queryparams_order": ["mangaid","plate","plateifu","name","z"],
              "returnparams": null,
              "runtime": {"days": 0,"microseconds": 55986,"seconds": 0},
              "totalcount": 4
           }

        '''
        searchfilter = args.pop('searchfilter', None)
        # searchfilter = self.results['inconfig'].get('searchfilter', None)
        # params = self.results['inconfig'].get('params', None)
        # rettype = self.results['inconfig'].get('returntype', None)
        # limit = self.results['inconfig'].get('limit', 100)
        # sort = self.results['inconfig'].get('sort', None)
        # order = self.results['inconfig'].get('order', 'asc')
        # release = self.results['inconfig'].get('release', None)

        try:
            # res = _getCubes(searchfilter, params=params, rettype=rettype,
            #                 limit=limit, sort=sort, order=order, release=release)
            res = _getCubes(searchfilter, **args)
        except MarvinError as e:
            self.results['error'] = str(e)
            self.results['traceback'] = get_traceback(asstring=True)
        else:
            self.results['status'] = 1
            self.update_results(res)

        # this needs to be json.dumps until sas-vm at Utah updates to 2.7.11
        return Response(json.dumps(self.results), mimetype='application/json')

    @route('/cubes/getcolumn/<colname>/', methods=['GET', 'POST'], endpoint='getcolumn')
    @av.check_args(use_params='query', required='searchfilter')
    def query_allcolumn(self, args, colname):
        ''' Retrieves the entire result set for a single column

        .. :quickref: Query; Retrieves the entire result set for a single column

        :query string release: the release of MaNGA
        :form searchfilter: your string searchfilter expression
        :resjson int status: status of response. 1 if good, -1 if bad.
        :resjson string error: error message, null if None
        :resjson json inconfig: json of incoming configuration
        :resjson json utahconfig: json of outcoming configuration
        :resjson string traceback: traceback of an error, null if None
        :resjson string data: dictionary of returned data
        :json list column: the list of results for the specified column
        :resheader Content-Type: application/json
        :statuscode 200: no error
        :statuscode 422: invalid input parameters

        **Example request**:

        .. sourcecode:: http

           GET /marvin2/api/query/cubes/getcolumn/cube.plateifu/ HTTP/1.1
           Host: api.sdss.org
           Accept: application/json, */*

        **Example response**:

        .. sourcecode:: http

           HTTP/1.1 200 OK
           Content-Type: application/json
           {
              "status": 1,
              "error": null,
              "inconfig": {"release": "MPL-5", "searchfilter": "nsa.z<0.1"},
              "utahconfig": {"release": "MPL-5", "mode": "local"},
              "traceback": null,
              "chunk": 100,
              "count": 4,
              "data": ["8485-1901", "8485-1902", "8485-12701", "7443-12701", "8485-12702"],
           }

        '''
        searchfilter = args.pop('searchfilter', None)
        format_type = args.pop('format_type', 'list')
        try:
            query, results = _run_query(searchfilter, **args)
        except MarvinError as e:
            self.results['error'] = str(e)
            self.results['traceback'] = get_traceback(asstring=True)
        else:
            self.results['status'] = 1
            if format_type == 'list':
                column = results.getListOf(colname, return_all=True)
            self.results['data'] = column

        return Response(json.dumps(self.results), mimetype='application/json')

    @route('/cubes/getsubset/', methods=['GET', 'POST'], endpoint='getsubset')
    @av.check_args(use_params='query', required=['searchfilter', 'start', 'end'])
    def query_getsubset(self, args):
        ''' Remotely grab a subset of results from a query

        .. :quickref: Query; Grab a subset of results from a remote query

        :query string release: the release of MaNGA
        :form searchfilter: your string searchfilter expression
        :form params: the list of return parameters
        :form rettype: the string indicating your Marvin Tool conversion object
        :form start: the starting page index of results you wish to grab
        :form end: the ending page index of the results you wish to grab
        :form limit: the limiting number of results to return for large results
        :form sort: a string parameter name to sort on
        :form order: the order of the sort, either ``desc`` or ``asc``
        :resjson int status: status of response. 1 if good, -1 if bad.
        :resjson string error: error message, null if None
        :resjson json inconfig: json of incoming configuration
        :resjson json utahconfig: json of outcoming configuration
        :resjson string traceback: traceback of an error, null if None
        :resjson string data: dictionary of returned data
        :json list results: the list of results
        :json string query: the raw SQL string of your query
        :json int chunk: the page limit of the results
        :json string filter: the searchfilter used
        :json list returnparams: the list of return parameters
        :json list params: the list of parameters used in the query
        :json list queryparams_order: the list of parameters used in the query
        :json dict runtime: a dictionary of query time (days, minutes, seconds)
        :json int totalcount: the total count of results
        :json int count: the count in the current page of results
        :resheader Content-Type: application/json
        :statuscode 200: no error
        :statuscode 422: invalid input parameters

        **Example request**:

        .. sourcecode:: http

           GET /marvin2/api/query/cubes/getsubset/ HTTP/1.1
           Host: api.sdss.org
           Accept: application/json, */*

        **Example response**:

        .. sourcecode:: http

           HTTP/1.1 200 OK
           Content-Type: application/json
           {
              "status": 1,
              "error": null,
              "inconfig": {"release": "MPL-5", "searchfilter": "nsa.z<0.1", "start":10, "end":15},
              "utahconfig": {"release": "MPL-5", "mode": "local"},
              "traceback": null,
              "chunk": 100,
              "count": 4,
              "data": [["1-209232",8485,"8485-1901","1901",0.0407447],
                       ["1-209113",8485,"8485-1902","1902",0.0378877],
                       ["1-209191",8485,"8485-12701","12701",0.0234253],
                       ["1-209151",8485,"8485-12702","12702",0.0185246]
                      ],
              "filter": "nsa.z<0.1",
              "params": ["cube.mangaid","cube.plate","cube.plateifu","ifu.name","nsa.z"],
              "query": "SELECT ... FROM ... WHERE ...",
              "queryparams_order": ["mangaid","plate","plateifu","name","z"],
              "returnparams": null,
              "runtime": {"days": 0,"microseconds": 55986,"seconds": 0},
              "totalcount": 4
           }

        '''
        searchfilter = args.pop('searchfilter', None)
        # searchfilter = self.results['inconfig'].get('searchfilter', None)
        # params = self.results['inconfig'].get('params', None)
        # start = self.results['inconfig'].get('start', None)
        # end = self.results['inconfig'].get('end', None)
        # rettype = self.results['inconfig'].get('returntype', None)
        # limit = self.results['inconfig'].get('limit', 100)
        # sort = self.results['inconfig'].get('sort', None)
        # order = self.results['inconfig'].get('order', 'asc')
        # release = self.results['inconfig'].get('release', None)

        try:
            # res = _getCubes(searchfilter, params=params, start=int(start),
            #                 end=int(end), rettype=rettype, limit=limit,
            #                 sort=sort, order=order, release=release)
            res = _getCubes(searchfilter, **args)
        except MarvinError as e:
            self.results['error'] = str(e)
            self.results['traceback'] = get_traceback(asstring=True)
        else:
            self.results['status'] = 1
            self.update_results(res)

        # this needs to be json.dumps until sas-vm at Utah updates to 2.7.11
        return Response(json.dumps(self.results), mimetype='application/json')

    @route('/getparamslist/', methods=['GET', 'POST'], endpoint='getparams')
    @av.check_args(use_params='query', required='paramdisplay')
    def getparamslist(self, args):
        ''' Retrieve a list of all available input parameters into the query

        .. :quickref: Query; Get a list of all or "best" queryable parameters

        :query string release: the release of MaNGA
        :form paramdisplay: ``all`` or ``best``, type of parameters to return
        :resjson int status: status of response. 1 if good, -1 if bad.
        :resjson string error: error message, null if None
        :resjson json inconfig: json of incoming configuration
        :resjson json utahconfig: json of outcoming configuration
        :resjson string traceback: traceback of an error, null if None
        :resjson string data: dictionary of returned data
        :json list params: the list of queryable parameters
        :resheader Content-Type: application/json
        :statuscode 200: no error
        :statuscode 422: invalid input parameters

        **Example request**:

        .. sourcecode:: http

           GET /marvin2/api/query/getparamslist/ HTTP/1.1
           Host: api.sdss.org
           Accept: application/json, */*

        **Example response**:

        .. sourcecode:: http

           HTTP/1.1 200 OK
           Content-Type: application/json
           {
              "status": 1,
              "error": null,
              "inconfig": {"release": "MPL-5"},
              "utahconfig": {"release": "MPL-5", "mode": "local"},
              "traceback": null,
              "data": ['nsa.z', 'cube.ra', 'cube.dec', ...]
           }

        '''
        paramdisplay = args.pop('paramdisplay', 'all')
        q = Query(mode='local', release=args['release'])
        if paramdisplay == 'all':
            params = q.get_available_params('all')
        elif paramdisplay == 'best':
            params = bestparams
        self.results['data'] = params
        self.results['status'] = 1
        output = jsonify(self.results)
        return output

    @route('/cleanup/', methods=['GET', 'POST'], endpoint='cleanupqueries')
    @av.check_args(use_params='query', required='task')
    def cleanup(self, args):
        ''' Clean up idle server-side queries or retrieve the list of them

        Do not use!

        .. :quickref: Query; Send a cleanup command to the server-side database

        :query string release: the release of MaNGA
        :form task: ``clean`` or ``getprocs``, the type of task to run
        :resjson int status: status of response. 1 if good, -1 if bad.
        :resjson string error: error message, null if None
        :resjson json inconfig: json of incoming configuration
        :resjson json utahconfig: json of outcoming configuration
        :resjson string traceback: traceback of an error, null if None
        :resjson string data: dictionary of returned data
        :json string clean: clean success message
        :json list procs: the list of processes currently running on the db
        :resheader Content-Type: application/json
        :statuscode 200: no error
        :statuscode 422: invalid input parameters

        **Example request**:

        .. sourcecode:: http

           GET /marvin2/api/query/cleanup/ HTTP/1.1
           Host: api.sdss.org
           Accept: application/json, */*

        **Example response**:

        .. sourcecode:: http

           HTTP/1.1 200 OK
           Content-Type: application/json
           {
              "status": 1,
              "error": null,
              "inconfig": {"release": "MPL-5"},
              "utahconfig": {"release": "MPL-5", "mode": "local"},
              "traceback": null,
              "data": 'clean success'
           }

        '''
        task = args.pop('task', None)
        if task == 'clean':
            q = Query(mode='local')
            q._cleanUpQueries()
            res = {'status': 1, 'data': 'clean success'}
        elif task == 'getprocs':
            q = Query(mode='local')
            procs = q._getIdleProcesses()
            procs = [{k: v for k, v in y.items()} for y in procs]
            res = {'status': 1, 'data': procs}
        else:
            res = {'status': -1, 'data': None, 'error': 'Task is None or not in [clean, getprocs]'}
        self.update_results(res)
        output = jsonify(self.results)
        return output
